pthread/pthread.c

```diff
@@ -183,6 +191,22 @@ static void pthread_full_key_cleanup(_pthread_key_cleanup_t *head)
 }
 
 
+static void _pthread_do_cleanup(pthread_ctx *ctx)
+{
+	pthread_cleanup_t *head = ctx->cleanup_list;
+	while (ctx->cleanup_list != NULL) {
+		void (*routine)(void *) = head->routine;
+		void *arg = head->arg;
+		ctx->cleanup_list = head->next;
+		free(head);
+		head = ctx->cleanup_list;
```

Comment:
imho better to do this after reacquiring the lock, cleanup_list head might've changed in the meantime